File: tree-walk-interpreter/go/internal/lox/lox_function.go
package lox

import (
	"fmt"
)

type loxFunction struct {
	declaration *Function
	closure     *environment
	initializer bool
}

func newLoxFunction(declaration *Function, closure *environment, initializer bool) *loxFunction {
	return &loxFunction{
		declaration: declaration,
		closure:     closure,
		initializer: initializer,
	}
}

func (f *loxFunction) arity() int {
	return len(f.declaration.Parameters)
}

func (f *loxFunction) call(interpreter *Interpreter, arguments []interface{}) (returnVal interface{}) {
	defer func() {
		if r := recover(); r != nil {
			if returnValue, ok := r.(returnControl); ok {
				if f.initializer {
					returnVal = f.closure.getAt(0, "this")
					return
				}

				returnVal = returnValue.value
				return
			}

			panic(r)
		}
	}()

	environment := newEnvironment(f.closure)
	for i, paramter := range f.declaration.Parameters {
		environment.define(paramter.Lexeme, arguments[i])
	}

	interpreter.executeBlock(f.declaration.Body, environment)
	if f.initializer {
		returnVal = f.closure.getAt(0, "this")
	}
	return
}

func (f *loxFunction) bind(instance *loxInstance) *loxFunction {
	environment := newEnvironment(f.closure)
	environment.define("this", instance)
	return newLoxFunction(f.declaration, environment, f.initializer)
}

func (f *loxFunction) String() string {
	return fmt.Sprintf("<fn %s>", f.declaration.Name.Lexeme)
}

File: tree-walk/cs/Lox/Interpreter/Callables/LoxFunction.cs
using System;
using Lox.Parser;

namespace Lox.Interpreter.Callables;

class LoxFunction : ILoxCallable
{
    private readonly Function declaration;
    private readonly Environment closure;
    private readonly bool isInitializer;

    public LoxFunction(Function declaration, Environment closure, bool isInitializer)
    {
        this.declaration = declaration;
        this.closure = closure;
        this.isInitializer = isInitializer;
    }

    public int Arity()
    {
        return declaration.Parameters.Count;
    }

    public object? Call(Interpreter interpreter, List<object> arguments)
    {
        var environment = new Environment(closure);
        for (var i = 0; i < declaration.Parameters.Count; i++)
        {
            environment.Define(declaration.Parameters[i].Lexeme, arguments[i]);
        }

        try
        {
            interpreter.ExecuteBlock(declaration.Body, environment);
        }
        catch (ReturnValue returnValue)
        {
            if (isInitializer)
            {
                return closure.GetAt(0, "this");
            }

            return returnValue.Value;
        }

        if (isInitializer)
        {
            return closure.GetAt(0, "this");
        }

        return null;
    }

    public LoxFunction Bind(LoxInstance instance)
    {
        var environment = new Environment(closure);
        environment.Define("this", instance);
        return new LoxFunction(declaration, environment, isInitializer);
    }

    public override string ToString()
    {
        return $"<fn {declaration.Name.Lexeme}>";
    }
}

